### PR TITLE
Music library artist roles node

### PR DESCRIPTION
**Pull request by DaveTBlake**
Add core processing to support a Roles node, similar to genre, that shows all the artist roles and navigates to the artists with that role. 

Krypton can process and store music file tag data about standard roles: composer, conductor, orchestra, lyricist, remixer, arranger, engineer, producer, DJmixer and mixer and any value pairs in the TMCL/performer tags e.g. drummer, piano, vocalist etc. giving the contribution an artist makes.

Currently there are default nodes for the standard roles, but users have to make custom nodes for the other roles they are interested in. Using custom nodes role filtering can be combined  with genre e.g. a node for classical music composers, or pop music drummers. That is flexible and powerful, but hard work for new users. A simple node showing all roles that have been scanned into the library is a useful addition.

It creates a new content type of "roles", Estuary modified to display this, Confulence seems to just work.

A node file with 
`<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>`
`<node order="1" type="filter" visible="Library.HasContent(Music)">`
 `<label>Roles content</label>`
 `<content>artists</content>`
 `<group>roles</group>`
`</node>`

or alternatively
`<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>`
`<node order="1" type="folder" visible="Library.HasContent(Music)">`
 `<label>Roles Path</label>`
`<path>musicdb://roles</path>`
`</node>`

If anyone woud like to try it out you need to rescan music files that have been sutably tagged too.

### CODE REVIEW

**Comment by DaveTBlake** [comment]
@Razzeee perhaps you could have a look at this, or anyone else give a new music thing a go?

**Comment by DaveTBlake** [comment]
Thanks @phil65 got this working in Estuary now.

I'm in a delemma about making this a default node with a menu xml file rather than a custom node. It is not a mainstream thing, many users may not want it or have their music tagged to populate it, and I don't want to clutter the UI. But without visibility users aren't going to know it is even possible. My initial offering was to create default role nodes for the standard roles e.g. composer, lyricist etc. under a fixed Roles menu. This still makes sense, but how does an dynamic all roles node fit in? Any thoughts?

I guess a way forward would be to get it working in the nightly, and some feedback from users.

**Comment by zag2me** [comment]
Get it in as default then people see the feature and maybe tag their files better.

If we hide it away then people are going to miss the feature.

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -3600,8 +3600,8 @@ bool CMusicDatabase::GetRolesNav(const std::string& strBaseDir, CFileItemList& i
       return false;
 
     // get roles with artists having that role
-    std::string strSQL = "SELECT idRole, strRole FROM role WHERE EXISTS "
-                         "(SELECT 1 FROM song_artist WHERE song_artist.idRole = role.idRole)";
+    std::string strSQL = "SELECT DISTINCT role.idRole, role.strRole FROM role "
+                         "JOIN song_artist ON song_artist.idRole = role.idRole ";
```
Comment (phate89) [inline xbmc/music/MusicDatabase.cpp]: Are the 2 queries the same? Because if they are (as I think) the exists might be faster...

Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Yes they are. When I tested them on SQLite the join was unexpectedly quicker than the exists, I guess it hits some index. But to be honest my testing was not extensive, and a different data combination could change things too, and MySQL could be different again. So I would go with the join, but it is pretty arbitary. But many thanks for the interest.

Comment (phate89) [inline xbmc/music/MusicDatabase.cpp]: I tested on my i7 (with fake data) and I got a better result in both Mysql and sqlite.. But I guess it's a quick query anyway so shouldn't be a big problem even if little bit slower..

Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: It's going to vary with data (mine was real but that is not definative), and yes it is a quick query any way so not a concern.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Pull request merged**